pkg/frontend/querymiddleware/labels_query_cache.go

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"fmt"
+	"math"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/cache"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"golang.org/x/exp/slices"
+
+	"github.com/grafana/mimir/pkg/util"
+)
+
+const (
+	labelNamesQueryCachePrefix  = "ln:"
+	labelValuesQueryCachePrefix = "lv:"
+
+	stringParamSeparator = rune(0)
+	stringValueSeparator = rune(1)
+)
+
+var (
+	prometheusMinTime = time.Unix(math.MinInt64/1000+62135596801, 0).UTC().UnixMilli()
+	prometheusMaxTime = time.Unix(math.MaxInt64/1000-62135596801, 999999999).UTC().UnixMilli()
```

Comment:
These look taken from here https://github.com/grafana/mimir/blob/989fd4d523c84324f9bd8f0072a10e52b06cc2e3/integration/util.go#L45-L51

And they're also used in `time_test.go`. WDYT about moving them to [time.go](https://github.com/grafana/mimir/blob/6030ca6446e0782d1e9e2adf4cb39a3cb9b26418/pkg/util/time.go) and exporting them?

Reply:
Good idea. Done here: https://github.com/grafana/mimir/pull/5426/commits/c3844b9e88a5c169439f58e54b49fa3229969b28

Reply:
Also opened a PR to expose them (if Prometheus maintainers are open to it) https://github.com/prometheus/prometheus/pull/12535 but I will not block on it.